src/network/p2p_topology.rs

```diff
@@ -1,22 +1,118 @@
 //! module defining the p2p topology management objects
 //!
 
+use bincode;
+use chain_core::property;
+use network_core::gossip;
 use poldercast::topology::{Cyclon, Module, Rings, Topology, Vicinity};
 use poldercast::Subscription;
 pub use poldercast::{Address, Id, InterestLevel, Node};
-use std::collections::BTreeMap;
-
-use std::sync::{Arc, RwLock};
+use std::{
+    collections::BTreeMap,
+    error, fmt,
+    io,
+    sync::{Arc, RwLock},
+    vec,
+};
 
 pub const NEW_TRANSACTIONS_TOPIC: u32 = 0u32;
 pub const NEW_BLOCKS_TOPIC: u32 = 1u32;
 
+#[derive(Debug, Clone)]
+pub struct Gossip(pub Vec<Node>);
+
+#[derive(Debug)]
+pub struct Error {
+    source: Box<dyn error::Error + Send + Sync>,
+}
+
+impl Error {
+    pub fn new<E>(source: E) -> Self
+    where
+        E: Into<Box<dyn error::Error + Send + Sync>>,
+    {
+        Error {
+            source: source.into(),
+        }
+    }
+}
+
+impl error::Error for Error {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        Some(self.source.as_ref())
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.source.fmt(f)
+    }
+}
+
+impl From<io::Error> for Error {
+    fn from(e:  io::Error) -> Self {
+        Error {
+            source: Box::new(e),
+        }
+    }
+}
+
+pub fn from_node_id(node_id: &gossip::NodeId) -> Id {
+    let bytes = node_id.to_bytes();
+    let mut buf = [0; 16];
+    buf[0..16].clone_from_slice(&bytes);
+    Id::from(u128::from_be_bytes(buf))
+}
+
+pub fn to_node_id(id: &Id) -> gossip::NodeId {
+    gossip::NodeId::from_slice(&id.as_u128().to_be_bytes()).unwrap()
+}
+
+impl gossip::Gossip for Gossip {
+    type NodeId = Id;
+    type Node = Node;
+
+    fn from_nodes<I>(iter: I) -> Self
+    where
+        I: IntoIterator<Item = Self::Node>,
+    {
+        Gossip(iter.into_iter().collect())
+    }
+}
+
+impl IntoIterator for Gossip {
+    type Item = Node;
+    type IntoIter = vec::IntoIter<Node>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
 /// object holding the P2pTopology of the Node
 #[derive(Clone)]
 pub struct P2pTopology {
     topology: Arc<RwLock<Topology>>,
 }
 
+impl property::Serialize for Gossip {
+    type Error = Error;
+
+    fn serialize<W: std::io::Write>(&self, writer: W) -> Result<(), Self::Error> {
+        bincode::serialize_into(writer, &self.0).map_err(|err| Error::new(err))
```

Comment:
Note that `Error` could just encapsulate the bincode error with no need for the box.